docs/arch/adr-35.md

```diff
@@ -0,0 +1,25 @@
+# ADR 35: Adding undo/redo to workflow steps
+
+8 November 2021
+
+## Context
+
+For Engaging Crowds, we wanted to support workflows with both branching steps (a single answer question leads to different tasks, depending on the selected answer) and recursive steps (a given task can be returned to and annotated multiple times for a single subject.) Volunteers should be able to move backward and forward through these workflows without losing any work they'd already done.
+
+## Decision
+
+We solved this problem by adding an undo/redo history stack to the classifier task area. [#2050](https://github.com/zooniverse/front-end-monorepo/pull/2050) added a global history stack (`store.annotatedSteps`). [#2520](https://github.com/zooniverse/front-end-monorepo/pull/2520) refines this by moving from a global history manager to individual history management for each subject (`subject.stepHistory`.)
+
+History is managed by the MST middleware [`UndoManager`](https://github.com/mobxjs/mobx-state-tree/blob/master/packages/mst-middlewares/README.md#undomanager), which records snapshots of a single subtree in the store. Each history item is a single step key and one or more annotation IDs, representing the current state of the task area. 'Back' loads the previous step key and annotation IDs from history. 'Next' loads the next step key and annotation IDs, or creates a new history item if one doesn't exist already. Going back then changing branches, for a branching workflow, clears any existing redo history and starts a fresh history from the current point.
+
+## Consequences
+
+- `workflow.configuration.persistAnnotations` defaults to `true` for all workflows. Annotation values are remembered unless we explicitly turn this feature off.
```

Comment:
Is there any reason to ever turn this off? Can we deprecate 'persist annotations' in the lab? Pinging @lcjohnso for this one.

Reply:
My knowledge about the "persist annotations" feature derives from [this Slack thread](https://zooniverse.slack.com/archives/C06LHUC6R/p1595437880226700), which points to [this PFE Github issue](https://github.com/zooniverse/Panoptes-Front-End/issues/5378).  In that thread, the take-away was: either use "persist annotations" or recursive workflows, but not both.  This was based on PFE implementations of these features.

Are these two features compatible with each other in FEM?  Because recursion is the only case where I think we'd want the ability to turn off persistence, but if they now play nice in FEM, then I don't know of any cases where "persist annotations" should be disabled.

Final disclaimer: I'm not the final word on this -- the issue should be raised among the FE team to make sure no one else knows of a case where persist annotations may need to be false.

Reply:
They work in the FEM. That’s what this ADR is about, in fact. 

HMS NHS is a live example that persists annotations across recursive tasks.

Reply:
Here’s the PR where we originally mentioned getting rid of it. https://github.com/zooniverse/front-end-monorepo/pull/2364

Reply:
Given this, seems OK to me.

Reply:
Cool, I've opened #2525 to remove it.

Reply:
Note to Shaun: this is a reminder that [FEM-Lab](https://github.com/zooniverse/Panoptes-Front-End/pull/6037) needs to have different behaviour from the existing PFE-Lab/Project Builder.

- On PFE-Lab, `workflow.annotation.persistAnnotations === undefined` implies _false_
- On FEM-Lab, `workflow.annotation.persistAnnotations === undefined` implies _true_